### ISSUE TITLE
TFServing supports collection of metrics with prometheus

### ISSUE THREAD

**Issue by jlewi** [kubeflow/kubeflow#1036]
We'd like TFServing to support exporting relevant metrics in a K8s/Prometheus compatible manner.

Here are some relevant ussues tracking the features in TF Serving

tensorflow/serving#462 - Server metrics
  * cfegly's [comment](https://github.com/tensorflow/serving/issues/462#issuecomment-367594255) is a good summary of what we would like
tensorflow/serving#800 - Collect servable metrics

**Comment by ynqa** [comment]
@jlewi FYI: I also implemented tf serving exporter: https://github.com/ynqa/tf_serving_exporter.

**Comment by ynqa** [comment]
@jlewi I also implemented tf serving exporter for prometheus: https://github.com/ynqa/tf_serving_exporter.

**Comment by jlewi** [comment]
@ynqa how is https://github.com/ynqa/tf_serving_exporter related to tensorflow/serving#800? Are both needed?

/cc @lluunn

**Comment by ynqa** [comment]
@jlewi Yes. In [tensorflow/serving#800](https://github.com/tensorflow/serving/pull/800), I implemented to enable to estimate request/error counts as metrics, and fetch them via grpc service `GetModelMetrics`.

So, the exporter with http endpoint like /metrics where prometheus pull them from is needed. `tf_serving` > (via grpc service) > `tf_serving_exporter` > `prometheus`.

However, recently tf serving is able to receive http requests, and it's maybe better to make the endpoint there directly for prometheus. But I'm not sure now.

**Comment by jlewi** [comment]
That's very helpful; thank you.

**Comment by bhack** [comment]
@jlewi What metrics do you like to export?

Cause some basic metrics for http could be collected also with https://github.com/prometheus/blackbox_exporter.

**Comment by jlewi** [comment]
@ynqa My sense is that it would be better if tensorflow/serving directly exported the metrics directly; just because its fewer missing pieces.

What do you think is the best next step? I'd like to target TF serving metrics collection for our 0.3 release end of Q.

Should we try to get the existing PR merged? Should we revise the existing PR to use the built in http server? Should we create a patched version of TFServing to test/experiment with?

It looks like the TFServing team never responded to tensorflow/serving#800 I'm happy to reach out internally to them; but it would be good to get clarity on what we think we(Kubeflow) want.

/cc @kunmingg

**Comment by ynqa** [comment]
@jlewi 

>  it would be better if tensorflow/serving directly exported the metrics directly

I couldn't agree more.

>Should we try to get the existing PR merged? 

Yes. If you estimates the metrics on the side of serving. As another way is to make the proxy (extension of tf_serving_exporter) in the front of tf serving requests, but I don't recommend.

> Should we revise the existing PR to use the built in http server?

Finally yes. It's the best that tf serving exported the metrics directly. First of all, it's good to prepare tf_serving_exporter in the front of serving, because REST is not sophisticated: it supports the prediction only (`GetModelStatus` is not supported), and is optional whether utilizing, [ref](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/model_servers/main.cc#L381-L384). 

> Should we create a patched version of TFServing to test/experiment with?

If tensorflow/serving#800 is merged, yes.

I could cooperate to resolve the issue. Thanks.

**Comment by jlewi** [comment]
@ynqa Do we want to update tensorflow/serving#800? We can then ping the TF Serving folks to get them to review and hopefully accept it.

**Comment by jlewi** [comment]
@ynqa Would you mind updating the PR just to resolve the conflicts?

Since getting code merged into TF Serving can be a slow process, I think it might be worthwhile to continue to push forward on the complete TF Serving monitoring story without being blocked on the PR being merged.

I think the goal would be to get the functionality incorporated into one of our examples (e.g. GitHub Issue Summarization) to the point where we could demo surfacing the relevant metrics in Prometheus.

What would be the other missing pieces? I think we would need to get the tf_serving_exporter added to the TF Serving deployment specs. Is there anything else we would need?

**Comment by ynqa** [comment]
@jlewi I just resolved the conflicts on PR.

Okay. The next step is:
1. It's good to enable to export all model metrics without setting model name. Now, `GetModelMetrics` is required a model name like `GetModelStatus` 
  - [model_service.proto](https://github.com/ynqa/serving/blob/metrics/tensorflow_serving/apis/model_service.proto#L14-L18) 
  - [tf_serving_exporter](https://github.com/ynqa/tf_serving_exporter/blob/master/tf_serving_exporter.py#L103-L108). 
  - tensorflow/serving#797.
1. adding the components into tf_serving_exporter (like tf_serving_model_state) to export predict/error count as metrics.
1. making the deployment manifest of tf_serving_exporter. Is the exporter executed as the sidecar of serving?

**Comment by jlewi** [comment]
Thanks @ynqa I'll ping some folks to try to get the PR reviewed.

**Comment by ynqa** [comment]
@jlewi Thanks too. Will it manage the exporter of serving in this org?

**Comment by jlewi** [comment]
@ynqa If we need additional logic/code/manifests beyond what's in TFServing repository then we can use kubeflow to host them. 

https://github.com/kubeflow/kubeflow/tree/master/components/k8s-model-server

is a good location to start putting code.

It would be great if you could start opening issues targeting 0.3 and submitting PRs.

**Comment by bhack** [comment]
@jlewi Are we interested also in gpu metrics? Can you check internally if we could have support directly in the node exporter https://github.com/prometheus/node_exporter/pull/755#issuecomment-403274849?

**Comment by jlewi** [comment]
@bhack yes we are also interested in GPU metrics. I'm not sure what the best story for those are. As this issue you linked to suggests some of these metrics should be collected at the node or possibly pod level. This issue is specifically about collecting app level metrics for TFServing. In that context, it would be up to TFServing to export GPU related metrics.

**Comment by bhack** [comment]
@jlewi Yes I really don't know what is best place. Probably, at node/pod level, could be useful also for training job and for monitoring/verify data pipeline efficiency that feeds the GPU. 

Cause at node level it is quite generic against TFServing/training or any other gpu activity. But probably we want something more specific.

P.s. that comment was by a GCP team member.

**Comment by ynqa** [comment]
@jlewi Sure, Thanks. BTW, when is the deadlines of 0.3?

**Comment by ynqa** [comment]
@jlewi There are the http proxy is already existed. I'll add the serving exporter here: https://github.com/kubeflow/kubeflow/tree/master/components/k8s-model-server/http-proxy

And also if tensorflow/serving#800 will be not merged, I try to make up metrics itself here. Is it okay?

**Comment by jlewi** [comment]
@ynqa That sounds good to me. 

We do quarterly releases. 0.3. will be end of September. So we aim to be code complete mid September.

**Comment by jlewi** [comment]
@lluunn What's the status of this? Are we blocked on TF releases?

**Comment by lluunn** [comment]
WIP, won't be in 0.3

**Comment by jlewi** [comment]
According to @lluunn #1911 added a prometheus endpoint to TFSeving. So only remaining piece is adding an annotation so that Prometheus knows to scrape it.
